src/inject.cc

```diff
@@ -8,68 +8,119 @@
 #include <tlhelp32.h>
 #include <vector>
 
-bool InjectDll(const uint32_t ProcessId, const std::filesystem::path &Path) {
+
+HANDLE OpenRemoteProcess(const uint32_t ProcessId) {
   const uint32_t ProcessRights =
       PROCESS_CREATE_THREAD | PROCESS_QUERY_INFORMATION | PROCESS_VM_OPERATION |
       PROCESS_VM_WRITE | PROCESS_VM_READ;
   const HANDLE Process = OpenProcess(ProcessRights, false, ProcessId);
 
   if (Process == nullptr) {
-    return EXIT_FAILURE;
+    printf("Failed to open the remote process.\n");
+    return INVALID_HANDLE_VALUE;
   }
 
-  const PVOID RemoteDllPath = VirtualAllocEx(
-      Process, nullptr, 0x1000, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+  return Process;
+}
+
+PVOID AllocateMemInRemoteProcess(HANDLE hRemoteProcess,
+                                 const std::filesystem::path &Path) {
+  const std::string DllPath = Path.string();
+  const size_t DllPathLen = DllPath.size() + 1;
+  SIZE_T BytesWritten;
+
+  const PVOID RemoteDllPath = VirtualAllocEx(hRemoteProcess, nullptr, 0x1000, MEM_COMMIT | MEM_RESERVE,
+                     PAGE_READWRITE);
 
   if (RemoteDllPath == nullptr) {
     printf("VirtualAllocEx failed.\n");
-    return EXIT_FAILURE;
+    goto Error;
   }
 
-  const std::string DllPath = Path.string();
-  const size_t DllPathLen = DllPath.size() + 1;
-  SIZE_T BytesWritten;
-  if (!WriteProcessMemory(Process, RemoteDllPath, DllPath.c_str(), DllPathLen,
+  if (!WriteProcessMemory(hRemoteProcess, RemoteDllPath, DllPath.c_str(),
+                          DllPathLen,
                           &BytesWritten)) {
+    VirtualFreeEx(hRemoteProcess, RemoteDllPath, 0, MEM_RELEASE);
     printf("WriteProcessMemory failed.\n");
-    return EXIT_FAILURE;
+    goto Error;
   }
 
+  return RemoteDllPath;
+
+Error:
+  CloseHandle(hRemoteProcess);
```

Comment:
nit: Can we rename this to `RemoteProcess`?

Reply:
I also wouldn't do any clean-up here but only in the caller; otherwise you end up with duplicated logic. It also saves you using `goto` in this function if you do that